### PR TITLE
feat(vtkVolumeRepresentationProxy): Disable volume rendering with 2D volumes

### PR DESCRIPTION
**Pull request by boucaud**
The volume mapper does not support rendering single-slice volumes. When loading these volumes with the vtkVolumeRepresentationProxy, this causes artifacts to be shown, but not the actual volume.

This is a problem in Paraview-Glance especially, as loading a scene exported from ParaView's slice representation doesn't show the expected result.

DeepinScreenshot_select-area_20201016204838

pvglance

This MR adds a tweak to vtkVolumeRepresentationProxy to make sure that we don't enable the volume mapper when rendering such volumes, but instead enable the slice actors.

Example volume: 

[singleSlice.vtkjs.zip](https://github.com/Kitware/vtk-js/files/5393173/singleSlice.vtkjs.zip)

### CODE REVIEW

**Comment by boucaud** [comment]
@finetjul

File: Sources/Proxy/Representations/VolumeRepresentationProxy/index.js [hunk 1]
```diff
@@ -222,18 +222,37 @@ function vtkVolumeRepresentationProxy(publicAPI, model) {
...
+    if (numberOfDimensions === 2) {
+      publicAPI.setHas2DVolume(true);
+    }
   }
 
   model.sourceDependencies.push({ setInputData });
 
   // API ----------------------------------------------------------------------
 
+  publicAPI.setHas2DVolume = (is2D) => {
```
Comment (finetjul) [inline Sources/Proxy/Representations/VolumeRepresentationProxy/index.js:239]: setHas2DVolume -> setIs2DVolume ?

File: Sources/Proxy/Representations/VolumeRepresentationProxy/index.js [hunk 2]
```diff
@@ -222,18 +222,37 @@ function vtkVolumeRepresentationProxy(publicAPI, model) {
...
+    model.has2DVolume = is2D;
+    if (is2D) {
+      model.volume.setVisibility(false);
+      publicAPI.setSliceVisibility(true);
```
Comment (finetjul) [inline Sources/Proxy/Representations/VolumeRepresentationProxy/index.js:243]: do not turn it on if it was not already on

File: Sources/Proxy/Representations/VolumeRepresentationProxy/index.js [hunk 3]
```diff
@@ -222,18 +222,37 @@ function vtkVolumeRepresentationProxy(publicAPI, model) {
...
+  };
+
   publicAPI.isVisible = () => model.volume.getVisibility();
 
   publicAPI.setVisibility = (isVisible) => {
     if (isVisible) {
-      // Turn on only the volume
-      model.volume.setVisibility(true);
+      if (model.has2DVolume) {
+        publicAPI.setSliceVisibility(true);
```
Comment (finetjul) [inline Sources/Proxy/Representations/VolumeRepresentationProxy/index.js:252]: Ideally it should enable only the appropriate slice

File: Sources/Proxy/Representations/VolumeRepresentationProxy/index.js [hunk 4]
```diff
@@ -222,18 +222,37 @@ function vtkVolumeRepresentationProxy(publicAPI, model) {
...
+      } else {
+        model.volume.setVisibility(true);
+      }
     } else {
       // Turn off everything
       model.volume.setVisibility(false);
```
Comment (finetjul) [inline Sources/Proxy/Representations/VolumeRepresentationProxy/index.js:258]:
```
model.volume.setVisibility(false);
publicAPI.setSliceVisibility(false);
```

Comment (finetjul) [inline Sources/Proxy/Representations/VolumeRepresentationProxy/index.js:258]: publicAPI.getVisibility should be 
```
publicAPI.getVisibility = () =>
    model.volume.getVisibility() ||
    publicAPI.getSliceVisibility();
```
and 
```
publicAPI.getSliceVisibility = model.actorX.getVisibility || model.actorY.getVisibility() ||
    model.actorZ.getVisibility();
```

File: Sources/Proxy/Representations/VolumeRepresentationProxy/index.js (hunk 2 above)
Comment (boucaud) [reply Sources/Proxy/Representations/VolumeRepresentationProxy/index.js:243]: The problem is that it's off by default on pv glance's 3D view, so we need to enable it somewhere.

Comment (finetjul) [reply Sources/Proxy/Representations/VolumeRepresentationProxy/index.js:243]: but 3D volume visibility is "true", no ?

Comment (boucaud) [reply Sources/Proxy/Representations/VolumeRepresentationProxy/index.js:243]: Yes, but we need slice visibility

File: Sources/Proxy/Representations/VolumeRepresentationProxy/index.js [hunk 5]
```diff
@@ -222,41 +222,66 @@ function vtkVolumeRepresentationProxy(publicAPI, model) {
...
   }
 
   model.sourceDependencies.push({ setInputData });
 
   // API ----------------------------------------------------------------------
 
+  publicAPI.setIs2DVolume = (is2D) => {
+    model.has2DVolume = is2D;
+    if (is2D) {
+      if (is2D && publicAPI.getVisibility()) {
```
Comment (finetjul) [inline Sources/Proxy/Representations/VolumeRepresentationProxy/index.js:242]: why check is2D again ?

File: Sources/Proxy/Representations/VolumeRepresentationProxy/index.js [hunk 6]
```diff
@@ -222,41 +222,66 @@ function vtkVolumeRepresentationProxy(publicAPI, model) {
...
   publicAPI.isVisible = publicAPI.getVisibility;
 
-  publicAPI.setSliceVisibility = macro.chain(
-    model.actorX.setVisibility,
-    model.actorY.setVisibility,
-    model.actorZ.setVisibility
-  );
-  publicAPI.getSliceVisibility = model.actorX.getVisibility;
+  publicAPI.setSliceVisibility = (isVisible) => {
+    if (isVisible && model.is2D) {
```
Comment (finetjul) [inline Sources/Proxy/Representations/VolumeRepresentationProxy/index.js:271]: you have inconsistent usage of `is2D` and `has2DVolume`

File: Sources/Proxy/Representations/VolumeRepresentationProxy/index.js [hunk 7]
```diff
@@ -222,41 +222,66 @@ function vtkVolumeRepresentationProxy(publicAPI, model) {
...
+      publicAPI.setIs2DVolume(true);
+    }
   }
 
   model.sourceDependencies.push({ setInputData });
 
   // API ----------------------------------------------------------------------
 
+  publicAPI.setIs2DVolume = (is2D) => {
+    model.has2DVolume = is2D;
```
Comment (finetjul) [inline Sources/Proxy/Representations/VolumeRepresentationProxy/index.js:240]: it is odd that you setIs2DVolume but the variable is named has2DVolume

**Pull request merged**